### ISSUE TITLE
elm-watch doesn't exit when stdin closes

### ISSUE THREAD

**Issue by chazsconi** [lydell/elm-watch#40]
In my use case I spawn `elm-watch hot` from my backend app which uses [Phoenix](https://www.phoenixframework.org/).  I configure Phoenix using a watcher (https://hexdocs.pm/phoenix/Phoenix.Endpoint.html#module-runtime-configuration).  I am also spawning `esbuild` for my other assets in the same way.

When Phoenix exits it signals to the child processes (`esbuild` and `elm-watch` in my case) that they should also exit by closing `stdin`, however, `elm-watch` doesn't exit and is thus left running, so needs to be killed manually.

This is easy to reproduce by running `npx elm-watch hot` and then pressing `ctrl-d` - this closes `stdin` but has no effect on `elm-watch` which continues running.

I think this should be the standard behaviour for tools like this as both `esbuild` and `webpack` terminate from their "watch" mode when `stdin` is closed.

I can work around this by wrapping `elm-watch` in this [script](https://hexdocs.pm/elixir/1.13/Port.html#module-zombie-operating-system-processes) but it would be nice if this was handled directly by `elm-watch`.

I tried to fix this, but my JS/Node knowledge is quite basic and I wasn't able to find the right place in the code to listen for `stdin` being closed and exit so unfortunately I cannot offer a PR.

**Comment by lydell** [comment]
Which elm-version are you using, btw?

**Comment by chazsconi** [comment]
0.19.1 on linux.

**Comment by lydell** [comment]
Oops, sorry, I meant elm-watch version 😅

**Comment by chazsconi** [comment]
I checked out `main` branch and built that.

**Comment by lydell** [comment]
I started looking into this, and I have some questions about Phoenix.

- Does Phoenix show the stdout/stderr of the watchers it runs? How? Is it mixed output from each watcher, a couple of lines from each?
- Is stdin connected to a TTY? Is stdout connected to a TTY? You can check by running `console.log({ "stdin.isTTY": process.stdin.isTTY, "stdout.isTTY": process.stdout.isTTY })` (put that in test.js and add `node test.js` as a watcher – it should log either `true` or `undefined` for each).
  - If _stdout_ is _not_ a TTY, elm-watch’s output turns into a log, where it never moves the cursor or clears the screen. If that’s the case, do you miss that?
  - If _stdin_ is _not_ a TTY, elm-watch won’t be able to get the color theme from your terminal to use in the error overlay (new in the beta).

Btw, `esbuild --watch` does not seem close when pressing ctrl+d either. But it does close if you put `esbuid --watch &` in test.bash and run `bash test.bash` (while elm-watch keeps running). So I’m going to follow that.

**Comment by lydell** [comment]
I started working on this in #42.

Is this a Phoenix-specific thing? I looks like it all started in https://github.com/brunch/brunch/issues/920 (by the creator of Elixir, and Brunch used to be like part of Phoenix or something, right?), and then frontend tool after frontend tool have added support, linking to other frontend tools doing it as reference, which always links back to https://github.com/brunch/brunch/issues/920 in the end.

To me, it sounds like a technical limitation of Erlang. https://github.com/brunch/brunch/issues/920 tries to justify it by many Unix tools working this way, and gives `cat` and `node` (not unix) as examples. Those two actively read stdin, so it makes sense. `sleep` however does not, and does not exit when stdin exits.

So, I’m not convinced “exit if stdin ends” is actually a thing. But if it’s easy to support, then why not? It all just feels a bit odd to me.

**Comment by lydell** [comment]
It looks like it’s not safe to change the behavior at this point, so I put it behind an environment variable in #42.

New idea: Is it possible to detect that we’re running as a Phoenix watcher? Then we could enable exit-on-stdin-end just for that.

**Comment by chazsconi** [comment]
Thanks a lot for all the effort you are doing for this!

> Does Phoenix show the stdout/stderr of the watchers it runs? How? Is it mixed output from each watcher, a couple of lines from each?

Yes - it mixes the output of all watchers and the output from Phoenix itself.  The output is in colour.

> Is stdin connected to a TTY? Is stdout connected to a TTY? You can check by running console.log({ "stdin.isTTY": process.stdin.isTTY, "stdout.isTTY": process.stdout.isTTY }) (put that in test.js and add node test.js as a watcher – it should log either true or undefined for each). 

This returns ` { 'stdin.isTTY': undefined, 'stdout.isTTY': undefined }`

Apologies - I was wrong that `esbuild --watch` does not exit on `ctrl-d`.  I realise now that it exits on `stdin` close because I use a custom `esbuild` script that I based on [this one](https://hexdocs.pm/phoenix/asset_management.html#esbuild-plugins) which has this in it:
```
process.stdin.on('close', () => {
                process.exit(0)
            })
```

Additionally, with `webpack` I now realise you need to set [this option](https://webpack.js.org/configuration/watch/#watchoptionsstdin) in the config or `--watch-stdin` on the cli for it to exit on when `stdin` closes.

> Is this a Phoenix-specific thing? I looks like it all started in https://github.com/brunch/brunch/issues/920 (by the creator of Elixir, and Brunch used to be like part of Phoenix or something, right?), and then frontend tool after frontend tool have added support, linking to other frontend tools doing it as reference, which always links back to https://github.com/brunch/brunch/issues/920 in the end.

You can use whatever bundler you want with Phoenix, but previously the recommendation in the docs and the tools that would generate a new Phoenix app template for you used `brunch`.  This then later changed to `webpack`, and then again to `esbuild`.

So I think you maybe right that this is a Phoenix specific thing due to an Erlang restriction, although I didn't investigate if the `--watch-stdin` feature was added to `webpack` just to support Phoenix.

Nevertheless it would be really nice to have this feature available in `elm-watch`. Trying to detect if it is running as a Phoenix watcher sounds a bit ugly and I'm not sure how you would do that in a robust future-proof way.  However having an environment variable or flag to enable exit on close of `stdin` sounds a good way to go.

**Comment by darioghilardi** [comment]
Hi, I also tested the fix using the env variable and it works fine, the config in Phoenix `dev.exs` is:
```
npx: [
      "elm-watch",
      "hot",
      cd: Path.expand("../assets", __DIR__),
      env: [{"ELM_WATCH_EXIT_ON_STDIN_END", "true"}]
    ],
```

@lydell and @chazsconi thanks a lot for taking care!
Could you release a new version of elm-watch please?

**Comment by lydell** [comment]
A new version coming soon! At least a beta. I was going to release it as non-beta, but found a (minor) regression.

**Comment by lydell** [comment]
Fixed the regression, released 1.1.0-beta.5 with this change. Hoping to release 1.1.0 the coming week.
